lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -478,17 +480,67 @@ class TraceProcessor {
   }
 
   /**
-   * @param {{candidateEventName: string, invalidateEventName: string, events: LH.TraceEvent[], timeOriginEvt: LH.TraceEvent}} options
-   * @return {{lcp: LH.TraceEvent | undefined, invalidated: boolean}}
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPEvent}
    */
-  static computeValidLCP(options) {
-    const {
-      candidateEventName,
-      invalidateEventName,
-      events,
-      timeOriginEvt,
-    } = options;
+  static isLCPEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Invalidate' &&
+        evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame);
+  }
+
+  /**
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPCandidateEvent}
+   */
+  static isLCPCandidateEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame && evt.args.data && evt.args.data.size !== undefined);
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} events
+   * @param {LH.TraceEvent} timeOriginEvent
+   * @return {LCPEvent | undefined}
+   */
+  static computeValidLCPAllFrames(events, timeOriginEvent) {
+    const lcpEvents = events.filter(this.isLCPEvent).reverse();
+
+    /** @type {Map<string, LCPCandidateEvent | undefined>} */
+    const lcpEventsByFrame = new Map();
+    for (const e of lcpEvents) {
+      if (e.ts <= timeOriginEvent.ts) break;
 
+      // Already found final LCP state of this frame.
+      const frame = e.args.frame;
+      if (lcpEventsByFrame.has(frame)) continue;
+
+      if (this.isLCPCandidateEvent(e)) {
+        lcpEventsByFrame.set(frame, e);
+      } else {
+        lcpEventsByFrame.set(frame, undefined);
```

Comment:
Comment that this is the invalidated event (right?)

Is this line even needed? seems that without this line, the next LCP candidate event gets picked up in the same manner.

Reply:
The line is needed so `lcpEventsByFrame.has()` will return true if the final event was an invalidate event. I will change the code to make this more clear, because it is a bit confusing.